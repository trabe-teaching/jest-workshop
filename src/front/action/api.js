const fetchOp = method => (url, opts) =>
  fetch(`http://localhost:9090${url}`, {
    ...opts,
    mode: "cors",
    method,
  }).then(
    res => {
      if (!res.ok) {
        throw new Error("Fetch error");
      }
      return res.json();
    },
    e => {
      throw e;
    },
  );

const get = fetchOp("get");
const post = fetchOp("post");
const put = fetchOp("put");

export const fetchUsers = () => get("/force-users");
export const seekMaster = id => put(`/seek/${id}`);
export const turn = id => put(`/turn/${id}`);
export const betray = id => put(`/betray/${id}`);
export const banish = id => put(`/banish/${id}`);
export const reset = () => post("/reset").then(fetchUsers);
