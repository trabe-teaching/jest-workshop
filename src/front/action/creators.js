import * as api from "action/api";

export const API_FAILED = "API_FAILED";

export const apiFailed = () => ({
  type: API_FAILED,
});

const fetchErrorHandler = dispatch => e => {
  console.error("Fetch failed!", e);
  dispatch(apiFailed());
};

export const USERS_UPDATED = "USERS_UPDATED";

export const usersUpdated = users => ({
  type: USERS_UPDATED,
  payload: {
    users,
  },
});

export const USERS_FETCH_REQUEST = "USERS_FETCH_REQUEST";
export const USERS_FETCH_SUCCESS = "USERS_FETCH_SUCCESS";

export const usersFetch = () => dispatch => {
  dispatch({ type: USERS_FETCH_REQUEST });

  return api
    .fetchUsers()
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: USERS_FETCH_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};

export const MASTER_SEEK_REQUEST = "MASTER_SEEK_REQUEST";
export const MASTER_SEEK_SUCCESS = "MASTER_SEEK_SUCCESS";

export const masterSeek = id => dispatch => {
  dispatch({ type: MASTER_SEEK_REQUEST });

  return api
    .seekMaster(id)
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: MASTER_SEEK_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};

export const TURN_REQUEST = "TURN_REQUEST";
export const TURN_SUCCESS = "TURN_SUCCESS";

export const turn = id => dispatch => {
  dispatch({ type: TURN_REQUEST });

  return api
    .turn(id)
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: TURN_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};

export const BETRAY_REQUEST = "BETRAY_REQUEST";
export const BETRAY_SUCCESS = "BETRAY_SUCCESS";

export const betray = id => dispatch => {
  dispatch({ type: BETRAY_REQUEST });

  return api
    .betray(id)
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: BETRAY_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};

export const BANISH_REQUEST = "BANISH_REQUEST";
export const BANISH_SUCCESS = "BANISH_SUCCESS";

export const banish = id => dispatch => {
  dispatch({ type: BANISH_REQUEST });

  return api
    .banish(id)
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: BANISH_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};

export const RESET_REQUEST = "RESET_REQUEST";
export const RESET_SUCCESS = "RESET_SUCCESS";

export const reset = () => dispatch => {
  dispatch({ type: RESET_REQUEST });

  return api
    .reset()
    .then(users => {
      dispatch(usersUpdated(users));
      dispatch({ type: RESET_SUCCESS });
    })
    .catch(fetchErrorHandler(dispatch));
};
