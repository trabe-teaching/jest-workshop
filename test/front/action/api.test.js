import { fetchUsers, seekMaster, reset } from "action/api";

global.fetch = (url, opts) => {
  switch (opts.method) {
    case "get":
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ fetch: true, url, opts }),
      });

    case "post":
      return Promise.resolve({
        ok: false,
      });

    default:
      return Promise.reject({ fetch: true, url, opts });
  }
};

describe("front/action/api", () => {
  describe("#fetchUsers", () => {
    it("GET /force-users", () => {
      expect(fetchUsers()).resolves.toMatchSnapshot();
    });
  });

  describe("#seekMaster", () => {
    it("PUT /seek/id", () => {
      expect(seekMaster(1)).rejects.toMatchSnapshot();
    });
  });

  describe("#reset", () => {
    it("POST /reset", () => {
      expect(reset()).rejects.toMatchSnapshot();
    });
  });
});
