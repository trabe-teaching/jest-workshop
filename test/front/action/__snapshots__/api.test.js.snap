// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`front/action/api #fetchUsers GET /force-users 1`] = `[get] http://localhost:9090/force-users {"mode":"cors"}`;

exports[`front/action/api #reset POST /reset 1`] = `[Error: Fetch error]`;

exports[`front/action/api #seekMaster PUT /seek/id 1`] = `[put] http://localhost:9090/seek/1 {"mode":"cors"}`;
